internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,443 @@
+	sess model.ExperimentSession,
+) {
+	for ctx.Err() == nil {
+		// read (timeout was set in Run)
+		buffer := make([]byte, 1024)
+		n, addr, err := pconn.ReadFrom(buffer)
+		respTime := time.Now()
+		if err != nil {
+			// stop if the connection is already closed
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			// store read failures and continue receiving
+			out <- responseInfo{t: respTime, err: err}
+			continue
+		}
+		resp := buffer[:n]
+
+		// dissect server response
+		supportedVersions, dst, err := m.DissectVersionNegotiation(resp)
+		if err != nil {
+			// the response was likely not the expected version negotiation response, so ignore it
+			sess.Logger().Infof(fmt.Sprintf("response dissection failed: %s", err))
+			continue
+		}
+		// propagate receive information
+		out <- responseInfo{raw: resp, t: respTime, dstID: hex.EncodeToString(dst), versions: supportedVersions}
+
+		sess.Logger().Infof("PING got response from %s", addr)
+	}
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	host := string(measurement.Input)
+	// allow URL input
+	if u, err := url.ParseRequestURI(host); err == nil {
+		host = u.Host
+	}
+	service := net.JoinHostPort(host, m.config.port())
+	udpAddr, err := net.ResolveUDPAddr("udp4", service)
+	if err != nil {
+		return err
+	}
+	rep := m.config.repetitions()
+	tk := &TestKeys{
+		Domain:      host,
+		Repetitions: rep,
+	}
+	measurement.TestKeys = tk
+
+	// create UDP socket
+	pconn, err := m.config.networkLibrary().ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return err
+	}
+	defer pconn.Close()
+
+	// set context and read timeouts
+	deadline := time.Duration(rep*5) * time.Second
+	pconn.SetReadDeadline(time.Now().Add(deadline))
+	ctx, cancel := context.WithTimeout(ctx, time.Second*time.Duration(5*rep))
+	defer cancel()
+
+	sendInfoChan := make(chan requestInfo)
+	recvInfoChan := make(chan responseInfo)
+	sendMap := make(map[string]*requestInfo)
+
+	// start sender and receiver goroutines
+	go m.sender(ctx, pconn, udpAddr, sendInfoChan, sess)
+	go m.receiver(ctx, pconn, recvInfoChan, sess)
+L:
+	for {
+		select {
+		case req := <-sendInfoChan: // a new ping was sent
+			if req.err != nil {
+				tk.Pings = append(tk.Pings, &SinglePing{
+					ConnIdDst:   req.dstID,
+					ConnIdSrc:   req.srcID,
+					Failure:     archival.NewFailure(req.err),
+					Request:     &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+					RequestTime: formatTime(req.t),
+				})
+				continue
+			}
+			sendMap[req.srcID] = &req
+
+		case resp := <-recvInfoChan: // a new response has been received
+			var (
+				req *requestInfo
+				ok  bool
+			)
+			// match response to request
+			if req, ok = sendMap[resp.dstID]; !ok {
+				continue // ignore any version negotiation response with an unknown destination ID
+			}
+			delete(sendMap, resp.dstID)
```

Comment:
I suggest to change the structure of the experiment as follows:

1. you allow more than a single response per ping
2. you keep the `sendMap` until the end
3. you store every reply into the `sendMap` by appending to the relevant ping
4. you include the case you are ignoring in the `receiver` in the results by returning an error
5. when you are done, you walk through the send map and assign `timeout` to every response-less entry
6. finally you transform the send map into a list of pings

This refactoring would allow us to detect multiple replies and spurious replies.